packages/start/src/server/collect-styles.ts

```diff
@@ -0,0 +1,166 @@
+      await findModuleDependencies(vite, module, ssr, deps);
+    }
+  }
+
+  async function addByUrl(url: string, ssr: boolean) {
+    const node = await getViteModuleNode(vite, url, ssr);
+
+    if (node) await add(node);
+  }
+
+  if (module.url.endsWith(".css")) return;
+
+  if (ssr) {
+    if (module.ssrTransformResult?.deps) {
+      for (const url of module.ssrTransformResult.deps) {
+        await addByUrl(url, ssr);
+      }
+
+      // Parallel version with incorrect style order
+      /* node.ssrTransformResult.deps.forEach((url) =>
+        branches.push(add_by_url(url, ssr)),
+      ); */
+    }
+  } else {
+    for (const { url } of module.importedModules) {
+      const node = await getViteModuleNode(vite, url, ssr);
+
+      if (node && !deps.has(node)) {
+        deps.add(node);
+        await findModuleDependencies(vite, node, ssr, deps);
+      }
+    }
+  }
+}
+
+
+// Vite doesn't expose these so we just copy the list for now
+// https://github.com/vitejs/vite/blob/d6bde8b03d433778aaed62afc2be0630c8131908/packages/vite/src/node/constants.ts#L49C23-L50
+const cssFileRegExp =
+  /\.(css|less|sass|scss|styl|stylus|pcss|postcss|sss)(?:$|\?)/;
+// https://github.com/vitejs/vite/blob/d6bde8b03d433778aaed62afc2be0630c8131908/packages/vite/src/node/plugins/css.ts#L160
+const cssModulesRegExp = new RegExp(`\\.module${cssFileRegExp.source}`);
+
+const isCssFile = (file: string) => cssFileRegExp.test(file);
+export const isCssModulesFile = (file: string) =>
+  cssModulesRegExp.test(file);
+
+// https://github.com/remix-run/remix/blob/65326e39099f3b2285d83aecfe734ba35f668396/packages/remix-dev/vite/styles.ts#L29
+const cssUrlParamsWithoutSideEffects = ["url", "inline", "raw", "inline-css"];
+export const isCssUrlWithoutSideEffects = (url: string) => {
+  const queryString = url.split("?")[1];
+
+  if (!queryString) {
+    return false;
+  }
+
+  const params = new URLSearchParams(queryString);
+  for (let paramWithoutSideEffects of cssUrlParamsWithoutSideEffects) {
+    if (
+      // Parameter is blank and not explicitly set, i.e. "?url", not "?url="
+      params.get(paramWithoutSideEffects) === "" &&
+      !url.includes(`?${paramWithoutSideEffects}=`) &&
+      !url.includes(`&${paramWithoutSideEffects}=`)
+    ) {
+      return true;
+    }
+  }
+
+  return false;
+};
+
+async function findFilesDepedencies(vite: ViteDevServer, files: Array<string>, ssr = false, deps = new Set<ModuleNode>()) {
+  for (const file of files) {
+    try {
+      const node = await getViteModuleNode(vite, file, ssr);
+      if (node) await findModuleDependencies(vite, node, ssr, deps);
+    } catch (e) {
+      console.error(e);
+    }
+  }
+
+  return deps;
+}
+
+const injectQuery = (url: string, query: string) =>
+  url.includes("?") ? url.replace("?", `?${query}&`) : `${url}?${query}`;
+
+export async function findStylesInModuleGraph(vite: ViteDevServer, id: string, ssr = false) {
+  const absolute = path.resolve(process.cwd(), id);
```

Comment:
I kinda recommend to use vite's resolveId here, as the id might be virtual 🤔. What do you think @Brendonovich?

E.g.:
```ts
const resolvedId = await vite.pluginContainer.resolveId(id, undefined, { ssr: false });
const absolute = resolvedId?.id ?? path.resolve(process.cwd(), id);
```